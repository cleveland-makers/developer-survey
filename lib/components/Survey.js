import React from 'react';
import PropTypes from 'prop-types';
import TellUsAboutYourCurrentRole from './pages/TellUsAboutYourCurrentRole';
import LetsTalkLanguages from './pages/LetsTalkLanguages';
import SpeakingOfWork from './pages/SpeakingOfWork';
import MovingOn from './pages/MovingOn';
import SomeOtherStuff from './pages/SomeOtherStuff';

let fieldValues = {
  name: null,
  email: null,
  password: null,
  age: null,
  colors: [],
};

const saveValues = fields => ((() => {
  fieldValues = Object.assign({}, fieldValues, fields);
})());

class Survey extends React.PureComponent {
  constructor(props) {
    super(props);
    this.updateState = this.updateState.bind(this);
  }

  state = {
    step: 1,
    currentRoles: [],
    yearsExp: 7,
  };

  updateState(state) {
    this.setState(Object.assign({}, this.state, state));
  }

  nextStep = () => {
    this.setState({
      step: this.state.step + 1,
    });
  }

  previousStep = () => {
    this.setState({
      step: this.state.step - 1,
    });
  }

  render() {
    switch (this.state.step) {
      case 1:
        return (<TellUsAboutYourCurrentRole
          nextStep={this.nextStep}
          saveValues={saveValues}
          {...this.props}
        />);
      case 2:
        return (<LetsTalkLanguages
          nextStep={this.nextStep}
          previousStep={this.previousStep}
          saveValues={saveValues}
          {...this.props}
        />);
      case 3:
        return (<SpeakingOfWork
          nextStep={this.nextStep}
          previousStep={this.previousStep}
          saveValues={saveValues}
          updateState={this.updateState}
          {...this.props}
        />);
      case 4:
        return (<MovingOn
          nextStep={this.nextStep}
          previousStep={this.previousStep}
          saveValues={saveValues}
          updateState={this.updateState}
          {...this.state}
          {...this.props}
        />);
      case 5:
        return (<SomeOtherStuff
          nextStep={this.nextStep}
          previousStep={this.previousStep}
          saveValues={saveValues}
          {...this.props}
        />);
      default:
        return (<TellUsAboutYourCurrentRole
          nextStep={this.nextStep}
          saveValues={saveValues}
          {...this.props}
        />);
    }
  }
}

Survey.propTypes = {
  survey: PropTypes.shape({
    careerDevelopmentJobCount: PropTypes.number.isRequired,
    careerSalary: PropTypes.number.isRequired,
    careerSatisfaction: PropTypes.string.isRequired,
    careerWorkLifeBalance: PropTypes.string.isRequired,
    developerCurrentRoles: PropTypes.arrayOf(PropTypes.string).isRequired,
    developerHowLong: PropTypes.number.isRequired,
    languagePrimaryHomeLanguages: PropTypes.arrayOf(PropTypes.string).isRequired,
    languagePrimaryWorkLanguage: PropTypes.string.isRequired,
    languageWhenDidYouLearnIt: PropTypes.number.isRequired,
    languageWhereDidYouLearnIt: PropTypes.arrayOf(PropTypes.string).isRequired,
    languageWhyDoYouUseIt: PropTypes.arrayOf(PropTypes.string).isRequired,
    officeEmployeeCount: PropTypes.string.isRequired,
    officeHoursPerWeek: PropTypes.number.isRequired,
    officeLocation: PropTypes.string.isRequired,
    personalEthnicity: PropTypes.string.isRequired,
    personalFavoriteClevelandActivity: PropTypes.string.isRequired,
    personalFavoriteSportsTeams: PropTypes.string.isRequired,
    personalGender: PropTypes.string.isRequired,
    personalHighestEducation: PropTypes.string.isRequired,
    personalWhyCleveland: PropTypes.arrayOf(PropTypes.string).isRequired,
  }).isRequired,
};

export default Survey;
