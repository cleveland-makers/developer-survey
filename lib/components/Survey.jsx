import React from 'react';
import PropTypes from 'prop-types';
import Diversity from './survey/Diversity';
import FourOhFour from './FourOhFour';
import LetsTalkLanguages from './survey/LetsTalkLanguages';
import MovingOn from './survey/MovingOn';
import SpeakingOfWork from './survey/SpeakingOfWork';
import SurveyProgress from './survey/SurveyProgress';
import TellUsAboutYourCurrentRole from './survey/TellUsAboutYourCurrentRole';

class Survey extends React.PureComponent {
  surveyPage() {
    switch (this.props.surveyStep) {
      case 0:
        return (<TellUsAboutYourCurrentRole
          {...this.props}
        />);
      case 1:
        return (<LetsTalkLanguages
          {...this.props}
        />);
      case 2:
        return (<SpeakingOfWork
          {...this.props}
        />);
      case 3:
        return (<MovingOn
          {...this.props}
        />);
      case 4:
        return (<Diversity
          {...this.props}
        />);
      default:
        return (<FourOhFour />);
    }
  }

  render() {
    return (
      <div>
        <SurveyProgress
          surveyStep={this.props.surveyStep}
          surveyLength={this.props.surveyLength}
        />
        {this.surveyPage()}
      </div>
    );
  }
}

Survey.propTypes = {
  surveyLength: PropTypes.number.isRequired,
  surveyStep: PropTypes.number.isRequired,
};

export default Survey;
