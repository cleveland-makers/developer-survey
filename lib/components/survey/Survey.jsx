import React from 'react';
import PropTypes from 'prop-types';
import Diversity from './Diversity';
import FourOhFour from '../FourOhFour';
import LetsTalkLanguages from './LetsTalkLanguages';
import Loading from '../Loading';
import MovingOn from './MovingOn';
import SpeakingOfWork from './SpeakingOfWork';
import storeProvider from '../storeProvider';
import SurveyComplete from '../SurveyComplete';
import SurveyNavigation from './SurveyNavigation';
import SurveyProgress from './SurveyProgress';
import TellUsAboutYourCurrentRole from './TellUsAboutYourCurrentRole';

function buttonOptions(currentStep) {
  if (currentStep === -1) {
    return {
      nextDisplay: false,
      previousDisplay: false,
    };
  } else if (currentStep === 0) {
    return {
      previousDisplay: false,
    };
  } else if (currentStep === 4) {
    return {
      nextHref: '/confirmation',
      nextLabel: 'Submit',
    };
  }
  return {};
}

function validator(survey, surveyStep) {
  return () => {
    switch (surveyStep) {
      case 0:
        return survey.developerCurrentRoles &&
          survey.developerCurrentRoles.length > 0 &&
          survey.developerHowLong !== 0;
      case 1:
        return survey.languagePrimaryHomeLanguages &&
          survey.languagePrimaryHomeLanguages.length > 0 &&
          survey.languagePrimaryWorkLanguage &&
          survey.languagePrimaryWorkLanguage.length > 0 &&
          survey.languageWhenDidYouLearnIt &&
          survey.languageWhenDidYouLearnIt > 0 &&
          survey.languageWhereDidYouLearnIt &&
          survey.languageWhereDidYouLearnIt.length > 0 &&
          survey.languageWhyDoYouUseIt &&
          survey.languageWhyDoYouUseIt.length > 0;
      case 2:
        return survey.careerDevelopmentJobCount > 0 &&
          survey.careerSalary > 0 &&
          survey.careerSatisfaction &&
          survey.careerSatisfaction.length > 0 &&
          survey.careerWorkLifeBalance &&
          survey.careerWorkLifeBalance.length > 0 &&
          survey.officeEmployeeCount &&
          survey.officeEmployeeCount.length > 0 &&
          survey.officeHoursPerWeek > 0 &&
          survey.officeLocation &&
          survey.officeLocation.length > 0;
      case 3:
        return survey.personalFavoriteClevelandActivity &&
          survey.personalFavoriteClevelandActivity.length > 0 &&
          survey.personalFavoriteSportsTeams &&
          survey.personalFavoriteSportsTeams.length > 0 &&
          survey.personalWhyCleveland &&
          survey.personalWhyCleveland.length > 0;
      case 4:
        return survey.personalEthnicity &&
          survey.personalEthnicity.length > 0 &&
          survey.personalGender &&
          survey.personalGender.length > 0 &&
          survey.personalHighestEducation &&
          survey.personalHighestEducation.length > 0;
      default:
        return false;
    }
  };
}

class Survey extends React.PureComponent {
  state = {
    loading: true,
  }

  componentDidMount = async () => {
    await this.props.store.loadProps();
    const { allowed } = await this.props.store.checkSurveyState();
    if (!allowed) {
      this.props.store.disallow();
    }
    this.props.store.resetState();
    this.setState({
      loading: false,
    });
  }

  surveyPage() {
    switch (this.props.surveyStep) {
      case -1:
        return (<SurveyComplete />);
      case 0:
        return (<TellUsAboutYourCurrentRole
          {...this.props}
        />);
      case 1:
        return (<LetsTalkLanguages
          {...this.props}
        />);
      case 2:
        return (<SpeakingOfWork
          {...this.props}
        />);
      case 3:
        return (<MovingOn
          {...this.props}
        />);
      case 4:
        return (<Diversity
          {...this.props}
        />);
      default:
        return (<FourOhFour />);
    }
  }

  render() {
    if (this.state.loading) {
      return (<Loading />);
    }
    const nextFunc = (this.props.surveyStep === 4) ?
      this.props.store.submitSurvey : null;
    const options = buttonOptions(this.props.surveyStep);
    const validation = validator(this.props.survey, this.props.surveyStep);
    return (
      <div>
        <SurveyProgress
          surveyStep={this.props.surveyStep}
          surveyLength={this.props.surveyLength}
        />
        <SurveyNavigation
          nextFunc={nextFunc}
          validator={validation}
          {...options}
        >
          {this.surveyPage()}
        </SurveyNavigation>
      </div>
    );
  }
}

Survey.propTypes = {
  store: PropTypes.shape({
    checkSurveyState: PropTypes.func.isRequired,
    disallow: PropTypes.func.isRequired,
    loadProps: PropTypes.func.isRequired,
    resetState: PropTypes.func.isRequired,
    submitSurvey: PropTypes.func.isRequired,
  }).isRequired,
  survey: PropTypes.shape({
    careerDevelopmentJobCount: PropTypes.number.isRequired,
    careerSalary: PropTypes.number.isRequired,
    careerSatisfaction: PropTypes.string.isRequired,
    careerWorkLifeBalance: PropTypes.string.isRequired,
    developerCurrentRoles: PropTypes.arrayOf(PropTypes.string).isRequired,
    developerHowLong: PropTypes.number.isRequired,
    languagePrimaryHomeLanguages: PropTypes.arrayOf(PropTypes.string).isRequired,
    languagePrimaryWorkLanguage: PropTypes.string.isRequired,
    languageWhenDidYouLearnIt: PropTypes.number.isRequired,
    languageWhereDidYouLearnIt: PropTypes.arrayOf(PropTypes.string).isRequired,
    languageWhyDoYouUseIt: PropTypes.arrayOf(PropTypes.string).isRequired,
    officeEmployeeCount: PropTypes.string.isRequired,
    officeHoursPerWeek: PropTypes.number.isRequired,
    officeLocation: PropTypes.string.isRequired,
    personalEthnicity: PropTypes.string.isRequired,
    personalFavoriteClevelandActivity: PropTypes.string.isRequired,
    personalFavoriteSportsTeams: PropTypes.string.isRequired,
    personalGender: PropTypes.string.isRequired,
    personalHighestEducation: PropTypes.string.isRequired,
    personalWhyCleveland: PropTypes.arrayOf(PropTypes.string).isRequired,
  }).isRequired,
  surveyLength: PropTypes.number.isRequired,
  surveyStep: PropTypes.number.isRequired,
};

export default storeProvider()(Survey);
