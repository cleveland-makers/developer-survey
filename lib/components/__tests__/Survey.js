import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Survey from '../Survey';

configure({ adapter: new Adapter() });

describe('Survey', () => {
  const testProps = {
    survey: {
      careerDevelopmentJobCount: 0,
      careerSalary: 0,
      careerSatisfaction: '',
      careerWorkLifeBalance: '',
      developerCurrentRoles: [],
      developerHowLong: 0,
      languagePrimaryHomeLanguages: [],
      languagePrimaryWorkLanguage: '',
      languageWhenDidYouLearnIt: 0,
      languageWhereDidYouLearnIt: [],
      languageWhyDoYouUseIt: [],
      officeEmployeeCount: '',
      officeHoursPerWeek: 0,
      officeLocation: '',
      personalEthnicity: '',
      personalFavoriteClevelandActivity: '',
      personalFavoriteSportsTeams: '',
      personalGender: '',
      personalHighestEducation: '',
      personalWhyCleveland: [],
    },
    surveyStep: 1,
    surveyLength: 5,
  };

  it('renders correctly', () => {
    const wrapper = shallow(<Survey
      {...testProps}
    />);

    expect(wrapper.find('SurveyProgress').length).toBe(1);
    expect(wrapper).toMatchSnapshot();
  });
});
