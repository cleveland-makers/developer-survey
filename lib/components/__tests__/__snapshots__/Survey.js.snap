// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flow renders TellUsAboutYourCurrentRole correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Survey
    surveyLength={0}
    surveyStep={0}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <SurveyProgress
          surveyLength={0}
          surveyStep={0}
        />,
        <TellUsAboutYourCurrentRoleContainer
          surveyLength={0}
          surveyStep={0}
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "surveyLength": 0,
          "surveyStep": 0,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "surveyLength": 0,
          "surveyStep": 0,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <SurveyProgress
            surveyLength={0}
            surveyStep={0}
          />,
          <TellUsAboutYourCurrentRoleContainer
            surveyLength={0}
            surveyStep={0}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "surveyLength": 0,
            "surveyStep": 0,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "surveyLength": 0,
            "surveyStep": 0,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
